src/vm_cpu.rs

```diff
@@ -526,6 +531,15 @@ impl<'vcpu, 'pages, 'host, T: GuestStagePagingMode> ActiveVmCpu<'vcpu, 'pages, '
     pub fn run(&mut self) -> VmCpuTrap {
         self.complete_pending_mmio_op();
 
+        // Set up our host's timer at HS level so that we trap if our host's timer expires.
+        if let VmCpuParent::HostVm(ref host_vcpu) = self.host_context {
+            let host_sie = LocalRegisterCopy::new(host_vcpu.vs_csrs().vsie);
+            if host_sie.read(sie::stimer) != 0 {
+                CSR.stimecmp.set(host_vcpu.vs_csrs().vstimecmp);
+                CSR.sie.read_and_set_field(sie::stimer);
+            }
```

Comment:
This will trigger timer before you actually want to. vstimecmp is time w.r.t host in VS mode.
However, stiemcmp should be programmed with vstimecmp + htdelta.

Reply:
We can also do a micro optimization here by first checking if the vstimecmp + htdelta > current_time.
In that case, the timer is already expired and host can be scheduled again.

Reply:
Ah good catch on `htimedelta`! Fixed. I've left the latter as a TODO for now while I think of a clean way to avoid duplicating the SGEI vs SEI detection logic.